lib/Dialect/Moore/MooreOps.cpp

```diff
@@ -310,6 +312,8 @@ LogicalResult VariableOp::canonicalize(VariableOp op,
 SmallVector<DestructurableMemorySlot> VariableOp::getDestructurableSlots() {
   if (isa<SVModuleOp>(getOperation()->getParentOp()))
     return {};
```

Comment:
It looks like we will tag global variables with struct type as destructurable. For example, struct values exist in `func.func`.

Reply:
I'm wondering if we may want to also allow for variables in `SVModuleOp` to be destructured by SROA. Would that be a problem?

Reply:
If we destructure in `Moore` dialect, mem2reg pass may eliminate part of struct variable, such as `%x.a` will be eliminated if it is unused. Some struct variable will be incomplete  which means struct information will lose in this IR.  Some optimizations for nested type may be disabled in lower IR.
What's more, most cases of struct variables operation will be transformed to non-struct variables operation unless this struct is used not by fielding. But in `hw dialect`, it still has many struct Ops. We do not need to destructure them in so higher IR.
I was concerning about this, wo I disabled SROA for global variables.

Reply:
There seems to be no problem. As the follow you mentioned, when `moore.read` exists, SROA has nothing to do, and vice versa. So I think maybe we can allow for variables in `SVModuleOp` to be destructured by SROA. Those variables with struct type don't expand by SROA, we can turn them into `moore.struct_create`(I notice canonicalize method can do this), then lower `moore.struct_create` directly into `hw.struct_create`. What do you think?

Reply:
Yeah that sounds great. We can't directly go from `moore.variable` to `moore.struct_create`, because the variable has `ref<struct>` type but `moore.struct_create` creates a `struct` directly. But in case Mem2Reg replaces the variable with SSA values, we can definitely materialize `moore.struct_create`. That would be especially nice also with `moore.assigned_variable` :smiley: